server/src/service.rs

```diff
@@ -107,10 +107,40 @@ impl Service {
     #[zbus(out_args("secrets"))]
     pub async fn get_secrets(
         &self,
-        _items: Vec<OwnedObjectPath>,
-        _session: ObjectPath<'_>,
+        items: Vec<OwnedObjectPath>,
+        session: OwnedObjectPath,
     ) -> Result<HashMap<OwnedObjectPath, SecretInner>, ServiceError> {
-        todo!()
+        let mut secrets = HashMap::new();
+        let collections = self.collections.lock().await;
+
+        'outer: for collection in collections.iter() {
+            for item in &items {
+                if let Some(item) = collection.get_item_from_path(item).await {
+                    match item.get_secret(session.clone()).await {
```

Comment:
Yeah I could have done this. But, actually we need the match. Because, consider a scenario where a correct Session is present/provided. But the item(s) are locked. In this situation we still need to return `secrets: HashMap<OwnedObjectPath, SecretInner>`. We can error out only if there is no session (returns nothing). Of course this is based on gnome-keyring behavior :)

Reply:
The match you wrote doesn't do what you are mentioning :)

Reply:
Ha :) yes, you're right.

The following is what we need:
Based on gnome-keyring behavior,  `Item.get_secret` call error out if there is no session or the item is locked. But `Service.get_secrets` only error out when there is no session.

I had this correctly implemented before [this](https://github.com/bilelmoussaoui/oo7/pull/153#discussion_r1844014981) change,

I added another arm:
```
Err(ServiceError::IsLocked(_)) => {                                                   
      continue;
 }
```

with this we don't error out anymore if the items are locked when calling Service.get_secrets.

If this is still bad, how can I improve this?

Reply:
Well, first of all this difference in behaviour between Item::get_secret & Service::get_secrets should be documented in Service::get_secrets along with a reference to gnome-keyring-daemon.

Also, you don't need all the other match statements, only the Ok/Err(IsLocked)/Err(e), are enough.

Reply:
Got it, thanks!